src/inmanta/agent/agent.py

```diff
@@ -1533,3 +1537,253 @@ async def get_facts(self, env: uuid.UUID, agent: str, resource: dict[str, Any])
+
+    :param agent: The Agent instance that this ExecutorManager is part of.
+    :param environment_manager: The VirtualEnvironmentManager responsible for managing the virtual environments
+    """
+
+    def __init__(self, agent: Agent, environment_manager: VirtualEnvironmentManager):
+        self.executor_map: dict[ExecutorId, Executor] = {}
+        self.environment_manager = environment_manager
+        self.agent = agent  # for now the executorManager lives in an agent
+        self.storage = self.create_storage()
+
+    async def create_executor(self, executor_id: ExecutorId) -> Executor:
+        """
+        Creates an Executor based with the specified agent name and blueprint.
+        It ensures the required virtual environment is prepared and source code is loaded.
+
+        :param executor_id: executor identifier containing an agent name and a blueprint configuration.
+        :return: An Executor instance
+        """
+        blueprint = executor_id.blueprint
+        env_blueprint = blueprint.to_env_blueprint()
+        venv = await self.environment_manager.get_environment(env_blueprint, self.agent.thread_pool)
+        executor = Executor(executor_id, venv, self.storage)
+        executor.load_code(blueprint.sources)
+        self.executor_map[executor_id] = executor
+        return executor
+
+    async def get_executor(self, agent_name: str, blueprint: ExecutorBlueprint) -> Executor:
+        """
+        Retrieves an Executor based on the agent name and blueprint.
+        If an Executor does not exist for the given configuration, a new one is created.
+
+        :param agent_name: The name of the agent for which an Executor is being retrieved or created.
+        :param blueprint: The ExecutorBlueprint defining the configuration for the Executor.
+        :return: An Executor instance
+        """
+        executor_id = ExecutorId(agent_name, blueprint)
+        if executor_id in self.executor_map:
+            return self.executor_map[executor_id]
+        return await self.create_executor(executor_id)
+
+    async def execute(
+        self,
+        agent_name: str,
+        blueprint: ExecutorBlueprint,
+        resources: list[Resource],
+    ) -> None:
+        """
+        Execute the given resources with the appropriate Executor.
+
+        :param agent_name: The name of the agent under which the execution is performed.
+        :param blueprint: The ExecutorBlueprint defining the configuration for the Executor.
+        :param resources: A list of Resource instances to be deployed.
+        """
+        executor = await self.get_executor(agent_name, blueprint)
+        executor.execute(resources)
+
+    def create_storage(self) -> str:
+        """
+        Prepares and returns the path to the storage directory used by Executors for their source code.
+
+        :return: The path to the storage directory.
+        """
+        state_dir = cfg.state_dir.get()
+        code_dir = os.path.join(state_dir, "codes")
```

Comment:
Do we have tickets to cleanup the old venv/code directories?

Reply:
https://app.zenhub.com/workspaces/core-5a7c152b04a1652024289b7b/issues/gh/inmanta/inmanta-core/7284